src/transformers/trainer_callback.py

```diff
@@ -475,3 +492,30 @@ def on_log(self, args, state, control, logs=None, **kwargs):
         _ = logs.pop("total_flos", None)
         if state.is_local_process_zero:
             print(logs)
+
+
+class EarlyStoppingCallback(TrainerCallback):
+    """
+    An early stopping callback
+    """
+
+    def check_metric_value(self, args, state, control, metric_value):
+        # best_metric is set by code for load_best_model
+        operator = np.greater if args.greater_is_better else np.less
+        if state.best_metric is None or (
+            operator(metric_value, state.best_metric)
+            and abs(metric_value - state.best_metric) > args.early_stopping_threshold
+        ):
+            state.early_stopping_patience_counter = 0
+        else:
+            state.early_stopping_patience_counter += 1
+
+    def on_best_metric_check(self, args, state, control, metric_value, **kwargs):
```

Comment:
Why can't this be done in `on_evaluate` directly?

Reply:
I can be done, and that's what I tried first, but if this is done in `on_evaluate` we'd have to duplicate the [logic to find the best_metric](https://github.com/huggingface/transformers/blob/b59e2650b6be1d1fe05197d2c8a5e13bc3a345b2/src/transformers/trainer.py#L910) in EarlyStoppingCallback. Personally, I like keeping the metric calculation synced but it's only a couple lines so I'll push something without that function synced and you can see if you like it.